tools/tdbg/dlq_v2_service.go

```diff
@@ -0,0 +1,179 @@
+	if c.IsSet(FlagLastMessageID) {
+		maxMessageID = c.Int64(FlagLastMessageID)
+		if maxMessageID < persistence.FirstQueueMessageID {
+			return fmt.Errorf(
+				"--%s must be at least %d but was %d",
+				FlagLastMessageID,
+				persistence.FirstQueueMessageID,
+				maxMessageID,
+			)
+		}
+	}
+
+	outputFile, err := getOutputFile(c.String(FlagOutputFilename))
+	if err != nil {
+		return err
+	}
+	adminClient := ac.clientFactory.AdminClient(c)
+
+	pageSize := c.Int(FlagPageSize)
+	iterator := collection.NewPagingIterator[*adminservice.HistoryDLQTask](
+		func(paginationToken []byte) ([]*adminservice.HistoryDLQTask, []byte, error) {
+			request := &adminservice.GetDLQTasksRequest{
+				DlqKey: &adminservice.HistoryDLQKey{
+					Category:      enums.TaskCategory(ac.category.ID()),
+					SourceCluster: ac.sourceCluster,
+					TargetCluster: ac.targetCluster,
+				},
+				PageSize:      int32(pageSize),
+				NextPageToken: paginationToken,
+			}
+			res, err := adminClient.GetDLQTasks(ctx, request)
+			if err != nil {
+				return nil, nil, fmt.Errorf("call to GetDLQTasks from ReadMessages failed: %w", err)
+			}
+			return res.DlqTasks, res.NextPageToken, nil
+		},
+	)
+
+	encoder := jsonpb.Marshaler{
+		Indent:       "  ",
+		EmitDefaults: true,
+	}
+	for iterator.HasNext() && remainingMessageCount > 0 {
+		dlqTask, err := iterator.Next()
+		if err != nil {
+			return fmt.Errorf("DLQ task iterator returned error: %w", err)
+		}
+		if dlqTask.Metadata.MessageId > maxMessageID {
+			break
+		}
+		remainingMessageCount--
+		// TODO: decode the task and print it in a human readable format
+		taskString, err := encoder.MarshalToString(dlqTask)
+		if err != nil {
+			return fmt.Errorf("unable to encode dlq message: %w", err)
+		}
+		_, err = outputFile.WriteString(fmt.Sprintf("%v\n", taskString))
+		if err != nil {
+			return fmt.Errorf("fail to print dlq messages: %s", err)
+		}
+	}
+	return nil
+}
+
+func (ac *DLQV2Service) PurgeMessages(*cli.Context) error {
+	return errors.New("purge is not yet implemented for DLQ v2")
+}
+
+func (ac *DLQV2Service) MergeMessages(*cli.Context) error {
+	return errors.New("merge is not yet implemented for DLQ v2")
+}
+
+func getSupportedDLQTaskCategories() []tasks.Category {
+	return []tasks.Category{
+		tasks.CategoryTransfer,
+		tasks.CategoryTimer,
+		tasks.CategoryReplication,
+		tasks.CategoryVisibility,
+		tasks.CategoryArchival,
+	}
+}
+
+func getCategoriesList() string {
+	var categoryString strings.Builder
+	categories := getSupportedDLQTaskCategories()
+	for i, c := range categories {
+		if i == len(categories)-1 {
+			categoryString.WriteString(" and ")
+		} else if i > 0 {
+			categoryString.WriteString(", ")
+		}
+		categoryString.WriteString(c.Name())
+	}
+	return categoryString.String()
+}
+
+func getCategoryByName(name string) (tasks.Category, bool) {
+	for _, c := range getSupportedDLQTaskCategories() {
+		if c.Name() == name {
+			return c, true
+		}
+	}
+	return tasks.Category{}, false
+}
```

Comment:
tdbg needs to work for queues not defined in this repo as well. I think we just need to pass server the categoryID, the flag desc should also reflect that (provide either the name or the ID). 
It's ok if tdbg doesn't know about the category, server will return QueueNotFound error if category ID is unknown to it.

Reply:
I want to fix this in a follow-up, which I already have a task for. Unfortunately, it's going to require quite a bit of changes. For example, we can't use the TaskCategory enum in our protos. We need to use a string type. I think we should have a task category registry option which can be provided to the fx graph of the history app, and validation can be done there. All clients of history (like admin service, the DLQ workflow and tdbg), will have to use string types.

Reply:
Ok, I made this use an integer instead of an enum or a string. Here’s why:
1. If it’s an enum, then, it would return unexpected values for `.String()` for dynamci task categories.
2. If it’s a string, then people might call the API with both enumspb.TASK_CATEGORY_TRANSFER.String()  and tasks.CategoryTransfer.Name(). However, only one would work because they have different values. The same goes for other categories.

If it’s an int, we don’t have either of those problems. Server code knows that it’s just an integer id, which they can look up using tasks.GetCategoryByID(…) and devs can call the API both with the enum and with the plain-old Go object’s ID.

In addition, I made the supported categories list include everything from `tasks.GetCategories()` except for memory-timer. This will support dynamic task categories, but it won't include the one non-dynamic category which we don't support for the DLQ.